### ISSUE TITLE
How to predict on one image?

### ISSUE THREAD

**Issue by EtagiBI** [qubvel/segmentation_models#338]
Hello,

I'm still trying to reproduce [this example](https://github.com/qubvel/segmentation_models/blob/master/examples/binary%20segmentation%20(camvid).ipynb) using Keras.

I was able to train the model, but I have problems with predicting. I want to see a prediction for a one single image. Here's my code:
```
    model.load_weights("camvid_model.h5")
    test_image = cv2.imread("Seq05VD_f01980.png") # It's an image from test folder of CamVid dataset
    ready_image = np.expand_dims(test_image, axis=0).astype('float')
    predicted_image = model.predict(ready_image)[0, :, :, :]
    predicted_image = (predicted_image * 255).astype(np.uint8)
    cv2.imwrite(f"prediction.jpg", predicted_image)
```
Here's a test image (there's a car in the center):
![Seq05VD_f01980](https://user-images.githubusercontent.com/20855725/82812119-b6b2f880-9eab-11ea-8c00-fcf16bcf5d72.png)

Here's my prediction for the test image:
![prediction](https://user-images.githubusercontent.com/20855725/82812132-bf0b3380-9eab-11ea-82ad-336fdf1dbf7a.jpg)

As you can see, my prediction is just a monochrome mess, not a car shape in the center. Are there any mistakes in my prediction code?

**Comment by bhomaidan1990** [comment]
@EtagiBI Your Question depends on the rest of your code `Training part`, and mask part, can you post a sample mask, and the training code, specially loss function, and metrics.

**Comment by EtagiBI** [comment]
@bhomaidan1990 my code is almost identcal to [this example](https://github.com/qubvel/segmentation_models/blob/master/examples/binary%20segmentation%20(camvid).ipynb). All functions and methods are untouched, so here's compilation ad training: 

```
BACKBONE = 'efficientnetb3'
IMAGE_SIZE = 320
BATCH_SIZE = 8
CLASSES = ['car']
LR = 0.0001
EPOCHS = 40
TRAIN = True

preprocess_input = sm.get_preprocessing(BACKBONE)


# define network parameters
n_classes = 1 if len(CLASSES) == 1 else (len(CLASSES) + 1)  # case for binary and multiclass segmentation
activation = 'sigmoid' if n_classes == 1 else 'softmax'

#create model
model = sm.Unet(BACKBONE, classes=n_classes, activation=activation)


# define optomizer
optim = keras.optimizers.Adam(LR)

# Segmentation models losses can be combined together by '+' and scaled by integer or float factor
dice_loss = sm.losses.DiceLoss()
focal_loss = sm.losses.BinaryFocalLoss() if n_classes == 1 else sm.losses.CategoricalFocalLoss()
total_loss = dice_loss + (1 * focal_loss)

# actulally total_loss can be imported directly from library, above example just show you how to manipulate with losses
# total_loss = sm.losses.binary_focal_dice_loss # or sm.losses.categorical_focal_dice_loss

metrics = [sm.metrics.IOUScore(threshold=0.5), sm.metrics.FScore(threshold=0.5)]

# compile keras model with defined optimozer, loss and metrics
model.compile(optim, total_loss, metrics)

# Dataset for train images
train_dataset = Dataset(
    x_train_dir,
    y_train_dir,
    classes=CLASSES,
    augmentation=get_training_augmentation(),
    preprocessing=get_preprocessing(preprocess_input),
)

# Dataset for validation images
valid_dataset = Dataset(
    x_valid_dir,
    y_valid_dir,
    classes=CLASSES,
    augmentation=get_validation_augmentation(),
    preprocessing=get_preprocessing(preprocess_input),
)

train_dataloader = Dataloder(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
valid_dataloader = Dataloder(valid_dataset, batch_size=BATCH_SIZE, shuffle=False)

# check shapes for errors
assert train_dataloader[0][0].shape == (BATCH_SIZE, IMAGE_SIZE, IMAGE_SIZE, 3)
assert train_dataloader[0][1].shape == (BATCH_SIZE, IMAGE_SIZE, IMAGE_SIZE, n_classes)

# define callbacks for learning rate scheduling and best checkpoints saving
callbacks = [
    keras.callbacks.ModelCheckpoint('./camvid_model.h5', save_weights_only=True, save_best_only=True, mode='min'),
    keras.callbacks.ReduceLROnPlateau(),
    keras.callbacks.EarlyStopping(patience=5, verbose=1)
]

if TRAIN:
    # train model
    history = model.fit_generator(
        train_dataloader,
        steps_per_epoch=len(train_dataloader),
        epochs=EPOCHS,
        callbacks=callbacks,
        validation_data=valid_dataloader,
        validation_steps=len(valid_dataloader),
    )
```

**Comment by JordanMakesMaps** [comment]
It doesn't look like you included the `preprocess_input` function in the prediction phase. The images you feed to the model during the prediction phase need to follow the exact same format that the images underwent during the training phase.

**Comment by EtagiBI** [comment]
@JordanMakesMaps does it mean that augmentation should also be applied to a test image?

**Comment by JordanMakesMaps** [comment]
@EtagiBI, nope, nor should it be used with the validation (just training). For validation and testing, just the pre-processing should be applied.

There is such as a thing called Test-Time Augmentation (TTA) which you might want to look into later down the road. It helps to provide slightly higher testing scores.

**Comment by EtagiBI** [comment]
Well, I was able to get a valid prediction using a `Dataset` instance with one single image inside. Not a very efficient way (and input processing in this case is quite complicated), but it works.
